### ISSUE TITLE
Custom behavior in tool mode when tilemap is edited from editor

### ISSUE THREAD

**Issue by tchernia** [godotengine/godot-proposals#668]
**Describe the project you are working on:**

Good day everybody! Thanks for that great game engine I enjoy using every day! :)

I am currently working on a game set in a post-apocalyptic world. The gameplay key is that the player can go back in time to solve puzzles! The base game is turn and tile based and we are using Godot tilemap features. Our kickstarter will be launched in one month and a half!

As we are making a professionnal game with lots of levels (roughly 60), we need a very good level design tool that ensures compliance of created maps and enforce constraints on created tiles. 

For example, I want to be able to put only one end tile on the map. Or I don't want to be able to put an obstacle on the cell where I put the player scene. It can seem stupid, but experience showed me that you actually make that kind of mistake even without knowing it (if offsets are not well fixed, or in case of other errors I cannot guess). 

So I much prefer securing those possibilities rather than trusting the level designer (even if it's me most of the time)

**Describe the problem or limitation you are having in your project:**

We are using the very powerful tool keyword for most of our level design functionalities. However as far as I know, there is no way to get a notification when tilemaps are updated, so it is impossible to secure tilemap changes.

I tried overriding set_cell and set_cellv functions, which would have been a great workaround, but it seems these functions are not called from the editor when changing a tile by hand (and I checked, I did not forget tool keyword on those files).

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

I would have loved one of these two features:

- Whether a notification or a signal can be intercepted when level designer erase or draws a tile on the tilemap.
- Or it is possible to override the function that is called when that happens.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

In GDScript:

```
func _on_tile_changed(Vector2 tile_position, int tile_id):

#Some code depending on your logic
```

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

Not as far as I know, because I can't get the information about the update.

**Is there a reason why this should be core and not an add-on in the asset library?:**

Exactly the same reason: if you can't get the information from the beginning you can't create a plugin doing it. Or I am missing something here, which is possible.

**Comment by samdze** [comment]
> I tried overriding set_cell and set_cellv functions, which would have been a great workaround, but it seems these functions are not called from the editor when changing a tile by hand (and I checked, I did not forget tool keyword on those files).

What do you mean by "changing a tile by hand"?

Are you sure this pull request doesn't address your issue? Overriding the set_cell method should be possible.
https://github.com/godotengine/godot/pull/27500

**Comment by tchernia** [comment]
Hello samdze!

"By hand" means by clicking left or right in the tilemap editor.

Yes it seems like the same issue! Good job! However I cannot use it in 3.1 (I did override set_cell but the function is not called from Tilemap editor when clicking), did you add the feature on the new version?

**Comment by samdze** [comment]
Yes, it is a Godot 3.2 feature, you can always upgrade your project (make a backup) and try again.
